src/VisualStudio/IntegrationTest/IntegrationTests/AbstractIdeIntegrationTest.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.IntegrationTest.Utilities;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.InProcess2;
+using Microsoft.VisualStudio.IntegrationTest.Utilities.Input;
+using Microsoft.VisualStudio.Threading;
+using Xunit;
+using ServiceProvider = Microsoft.VisualStudio.Shell.ServiceProvider;
+
+namespace Roslyn.VisualStudio.IntegrationTests
+{
+    [CaptureTestName]
+    public abstract class AbstractIdeIntegrationTest : IAsyncLifetime, IDisposable
+    {
+        protected readonly string ProjectName = "TestProj";
+        protected readonly string SolutionName = "TestSolution";
+
+        private JoinableTaskContext _joinableTaskContext;
+        private JoinableTaskCollection _joinableTaskCollection;
+        private JoinableTaskFactory _joinableTaskFactory;
+
+        private CancellationTokenSource _hangMitigatingCancellationTokenSource;
+
+        protected AbstractIdeIntegrationTest()
+        {
+            var componentModel = (IComponentModel)ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel));
+            JoinableTaskContext = componentModel.GetExtensions<JoinableTaskContext>().SingleOrDefault() ?? new JoinableTaskContext();
```

Comment:
For Roslyn, I could change this to a direct reference to `ThreadHelper.JoinableTaskContext`. For the generalized test harness, we'll want to make sure the "correct" instance is used whenever possible, and one is simply created for the other cases.

Reply:
So for the general case, you'd want to compile against vs-threading v12.x then, and let the binding redirect work in v15 so that you'd run downlevel properly.
But also, I wonder if JTF is really buying you anything over simply using the Dispatcher here. If you never block the UI thread as a test harness, I don't see what JTF is adding.